pyphare/pyphare/pharesee/geometry.py

```diff
@@ -0,0 +1,277 @@
+from ..core import box as boxm
+from .hierarchy import FieldData, is_root_lvl
+
+
+
+def toFieldBox(box, patch_data):
+    """
+    grows the box by one cell if patch_data is primal
+    because box.upper+1 allows to get the rightmost primal node
+    otherwise missing
+    """
+    assert (isinstance(patch_data, FieldData))
+    if patch_data.layout.centering["X"][patch_data.field_name] == "primal":
+        return boxm.Box(box.lower, box.upper + 1)
+    else:
+        return box
+
+
+
+
+
+def shift_patch(patch, offset):
+    patch.box = boxm.shift(patch.box, offset)
+    for pdata in patch.patch_datas.values():
+        pdata.ghost_box = boxm.shift(pdata.ghost_box, offset)
+
+
+
+
+def touch_domain_border(box, domain_box, border):
+    if border == "upper":
+        return domain_box.upper in box
+    elif border == "lower":
+        return domain_box.lower in box
+    else:
+        raise RuntimeError("invalid border")
+
+
+
+def compute_overlaps(patches, domain_box):
+    """
+    returns a list of overlaps for all patch datas in given patches
+    and for a domain box. An overlap is defined as an intersection of
+    two PatchData ghost box.
+
+    if  - = domain cell, o = ghost cell
+
+    PatchData1  : - - - - - - - - - o o o o o
+    PatchData2  :                     o o o o o - - - - - - - -
+    overlap box :                     ^ ^ ^ ^
+
+    for now the function is purely 1D and assumes the direction is
+    periodic. Therefore, PatchData close to the lower (resp. upper)
+    domain boundary (hence the need of domain_box) may overlap another
+    one close to the upper (resp. lower) end of the domain.
+
+    an overlap is a dictionnary with 3 keys:
+    - pdatas : value is a tuple of the two overlapped PatchDatas.
+               this is useful when one wants to get overlapped data
+    - box    :  is the intersection box between the two PatchData ghost boxes
+    - offset : tuple of two offsets by which the patchData ghost box is shifted
```

Comment:
right here, it is not obvious to me why it is 2 offsets while we are in 1D.

Reply:
because there are always two patchdata concerned by one overlap, and the box is defined by two AMR indexed points lower/upper which could need to be offseted to intersect one of the patchdata... since we don't want to think about which one it's safer to have two offsets and systematically apply an offset.

Reply:
ok, so I would suggest the comment
- offset : tuple of two offsets by which each patchData ghost box is shifted